### DISCUSSION TITLE
Only upgrade a specific package

### DISCUSSION THREAD

**Discussion by amskr** [rickysarraf/apt-offline#156] [Q&A] [answered]
Hi,

I was wondering if it's possible, or could easily be made possible, to create an upgrade for a specific package rather than including everything that needs upgrading in a bundle.

After creating a bundle using "apt-offline get upgrade.sig --bundle upgrade.zip" and then adding it to the cache of the offline macnine, I could run "apt-get install --only-upgrade <packagename>" to only upgrade the package I'm interested in but I would have had to put a zip including upgrades for everything on the offline machine to do this.

I'm essentially trying to avoid having to download a 300mb zip file when I only need 5mb of the data in it for a specific package as my offline machines are not truly offline but are rather on slow, less reliable connections with very restrictive data usage limits. Apt-offline is a great solution for installing new packages on them but not updating packages unless I'm missing something.

Would it be possible to add an option to apt-offline, maybe in the format "sudo apt-offline set package.sig --upgrade-packages <packagename>" such that the corresponding "apt-offline get package.sig --bundle package.zip" generates a package.zip that only includes the packages required to upgrade the named package?

Thanks for your time in advance

**Comment by rickysarraf** [comment, marked as answer]
That functionality is already there. Just the terminology is different.

```
rrs@priyasi:~$ apt-offline set -h
usage: apt-offline set [-h] [--verbose] [--simulate] [--install-packages [PKG ...]] [--install-src-packages [SOURCE PKG ...]] [--src-build-dep]
                       [--release release_name] [--update] [--upgrade] [--upgrade-type upgrade] [--generate-changelog] [--apt-backend apt-get]
                       apt-offline.sig

positional arguments:
  apt-offline.sig       Generate a signature file

optional arguments:
  -h, --help            show this help message and exit
  --verbose             Enable verbose messages
  --simulate            Just simulate. Very helpful when debugging
  --install-packages [PKG ...]
                        Packages that need to be installed
```

You can use the `--install-packages` option in the very same way you describe your use case.

**Reply by amskr** [reply]
Hi,

Thanks for the quick reply. Would you be able to provide me with an example as I think I must be missing how to do it.

If for example I want to update PHP on my offline machine, I've checked with "apt list –upgradeable" that updates are available and then run "sudo apt-offline set php-offline.sig --install-packages php" to create the signature file. On the online machine I then run "apt-offline get php-offline.sig --bundle php-install.zip" and it produces a 0kb zip file (presumably because PHP is already installed on the offline machine).

If I run "sudo apt-offline set php-offline.sig --upgrade --install-packages php" on the offline machine and then "apt-offline get php-offline.sig --bundle php-install.zip" on the online machine, it produces a 300mb zip file including lots of packages that are nothing to do with the PHP update or its dependencies when I only want the files related to the PHP update and its dependencies to be included.

Hopefully I'm just being a bit dumb here and an example can help clear things up.

Thanks

**Reply by rickysarraf** [reply]
Here's a real example.

```
rrs@priyasi:~$ apt policy hugo
hugo:
  Installed: 0.82.1-1
  Candidate: 0.82.1-1
  Version table:
     0.84.2-1 500
        500 http://deb.debian.org/debian unstable/main amd64 Packages
 *** 0.82.1-1 990
        990 http://deb.debian.org/debian testing/main amd64 Packages
        100 /var/lib/dpkg/status
     0.80.0-6+b5 500
        500 http://deb.debian.org/debian bullseye/main amd64 Packages
     0.55.6+really0.54.0-1 500
        500 http://deb.debian.org/debian buster/main amd64 Packages
20:04 ♒ ॐ ♅ ♄ ⛢     ☺ 😄    

rrs@priyasi:~$ sudo apt-offline set /tmp/hugo.uris --install-packages hugo --release unstable
Gathering installation details for package ['hugo']
20:05 ♒ ॐ ♅ ♄ ⛢     ☺ 😄    

rrs@priyasi:~$ \cat /tmp/hugo.uris 
'http://deb.debian.org/debian/pool/main/h/hugo/hugo_0.84.2-1_amd64.deb' hugo_0.84.2-1_amd64.deb 9696764 MD5Sum:e27107bf670326a37ee4e50453c2ea45
20:06 ♒ ॐ ♅ ♄ ⛢     ☺ 😄    
```

**Reply by amskr** [reply]
Thanks, I’m on holiday for a week now but will try it when I get back and let you know if I have any issues.

Thanks again

**Reply by amskr** [reply]
I've now tried based on your example and got it working.

Thanks for your help :-)
